H2 pollset impr

icing opened the pull request:
* `core`: add hook `create_secondary_connection` and method to core, since the implementation details for those connections to a master one belong there.
 *  `mod_http2`: use new hook and also recycle transit pools/bucket_alloc for those connections.
 *  `mod_http2`: improvements in pollset handling for secondary connections.

Commit:
*) mod_http2: removing the in_drain pipe and always doing the
     handling of stream input consumed as in the non-pollset version.
     Eliminating the special mplx_pool, basing pipes solely on the
     c2->pool which seems to work fine.

Commit:
*) mod_http2: unified c2 connection abort. making all
     volatiles into atomic operations instead.

Commit:
*) core: add new hook 'create_secondary_connection` and method
     `ap_create_secondary_connection()` to the core for setting
     up http2-like connections based on a master one. This brings
     implementation details to where they should be.
     mod_http2: use new ap_create_secondary_connection() from core
     and also recycle the transit pools/bucket_alloc used for those.

ylavic commented on modules/http2/h2_conn_ctx.h:
```diff
@@ -48,20 +50,17 @@ struct h2_conn_ctx_t {
-    volatile int done;               /* c2: processing has finished */
+    apr_uint32_t started;            /* c2: processing was started */
     apr_time_t started_at;           /* c2: when processing started */
+    apr_uint32_t done;               /* c2: processing has finished */
```
Maybe leave the `volatile` indicator for `started` and `done` to document that they are accessed atomically (now explicitely) still?

icing replied:
good idea, will add the `/* volatile */` comments. *scratch*

What about adding `/* atomic */`? Seems even more clear.

ylavic replied:
+1

icing replied:
added

ylavic commented on modules/http2/h2_mplx.c:
```diff
@@ -173,6 +176,63 @@ static void m_stream_cleanup(h2_mplx *m, h2_stream *stream)
+     * another thread.
+     */
+
+    apr_allocator_create(&allocator);
```
This could fail on OOM, `return NULL` or simply `ap_abort_on_oom()` in this case?
Since allocation failures on ptrans will abort() I wonder if allocating ptrans itself shouldn't abort() too, thus here and in case of `apr_pool_create_ex()` below.
This'd make checking `c2_transit_{create,get}()` for NULL is the caller(s) moot.

icing replied:
Added the checks. Thanks.

ylavic commented on modules/http2/h2_mplx.c:
```diff
@@ -254,18 +314,20 @@ h2_mplx *h2_mplx_c1_create(h2_stream *stream0, server_rec *s, apr_pool_t *parent
                                      m->pool);
     if (APR_SUCCESS != status) goto failure;
-    m->streams_input_read = h2_iq_create(m->pool, 10);
+#if !H2_POLL_STREAMS
```
Can't we do the same thing for output streaming, i.e. wakeup pollset instead of pipe putc in `c2_beam_output_write_notify()`?
And then get rid of `H2_POLL_STREAMS`, but I may be missing something..

icing replied:
I thought about this. We have support for disabling this bc Windows and older APR versions.

The possible benefit of the file descriptors is that we might adapt `mod_proxy` modules to use them for shuffling content back and forth, as you did for the wstunnel.

But then again, as long as we do not have that on all platforms we support, this means different code paths and complexity.

ylavic replied:
Oh, I wasn't thinking of removing file descriptors but wondered how you managed to remove the `H2_POLL_STREAMS` support for `streams_input_read` and why you couldn't do the same for the output side.

ylavic commented on modules/http2/h2_mplx.c:
```diff
@@ -254,18 +314,20 @@ h2_mplx *h2_mplx_c1_create(h2_stream *stream0, server_rec *s, apr_pool_t *parent
     mplx_pollset_add(m, conn_ctx);
 
     m->scratch_r = apr_pcalloc(m->pool, sizeof(*m->scratch_r));
+    m->max_spare_transits = 3;
```
Magic value `3` => `#define` or setting?
Also `m->max_spare_transits` is the same as `m->c2_transits->nalloc` so you might want to use the latter only all over the code?

icing replied:
This was a start setting and I plan to continue work on this.

Atm I am considering a global reserve list, because when connection go idle, this just hangs around useless.
The global list should probably keep not more than H2MaxWorker transits around.

ylavic commented on modules/http2/h2_mplx.c:
```diff
@@ -896,8 +940,8 @@ static void s_c2_done(h2_mplx *m, conn_rec *c2, h2_conn_ctx_t *conn_ctx)
-    ap_assert(conn_ctx->done == 0);
-    conn_ctx->done = 1;
+    AP_DEBUG_ASSERT(apr_atomic_read32(&conn_ctx->done) == 0);
+    apr_atomic_set32(&conn_ctx->done, 1);
```
TOCTOU, should this be:
```
    if (apr_atomic_cas32(&conn_ctx->done, 1, 0) != 0) {
        APR_DEBUG_ASSERT(0);
    }
```
?
And maybe `return;` if already set too? It looks like `s_c2_done()` is not supposed to be called multiple times and if it happens httpd's teeth will fall out (or can we simply ignore it)?

ylavic replied:
Oh this is the only place where `->done` is set (so not a TOCTOU issue, no CAS needed)..
Possibly `return;` on already set is meaningful still, or it should be an `ap_assert()`?

icing replied:
A DEBUG_ASSERT() seems best here. It is really only a dev check if things got screwed up by a change. This method needs to be called once and only once for a c2 connection.

ylavic replied:
Fair enough.

ylavic commented on modules/http2/h2_workers.c:
```diff
@@ -40,7 +40,7 @@ struct h2_slot {
     apr_thread_mutex_t *lock;
     apr_thread_cond_t *not_idle;
-    volatile apr_uint32_t timed_out;
+    apr_uint32_t timed_out;
```
`volative` (or `/* volatile */`) might be documentary still

icing replied:
ack

icing replied:
added

ylavic commented on modules/http2/h2_workers.c:
```diff
@@ -127,10 +127,9 @@ static void wake_idle_worker(h2_workers *workers)
     h2_slot *slot = pop_slot(&workers->idle);
     if (slot) {
-        int timed_out = 0;
+        int timed_out;
```
It looks like `wake_idle_worker()` could use a loop like:
```
    for (;;) {
        slot = pop_slot());
        if (!slot) {
            if (workers->dynamic && ...) { 
                add_worker(workers);
            }
            return;
        }
        if (!apr_atomic_read32(&slot->timed_out)) {
            lock();
            signal();
            unlock();
            return;
        }
        slot_done();
    }
```
rather than a recursion?

icing replied:
Changed as suggested.

ylavic commented on modules/http2/h2_workers.h:
```diff
@@ -38,19 +38,17 @@ struct h2_workers {
+    apr_uint32_t worker_count;
+    apr_uint32_t max_idle_secs;
+    apr_uint32_t aborted;
+    apr_uint32_t shutdown;
```
Ditto for volatile "documentation" (for those which are still accessed atomically)

icing replied:
ack.

ylavic commented on server/connection.c:
```diff
@@ -66,6 +70,13 @@ AP_DECLARE(conn_rec *) ap_create_connection(apr_pool_t *p,
+                                                      apr_bucket_alloc_t *alloc)
+{
+    return ap_run_create_secondary_connection(p, master, alloc);
+}
```
I'm not sure we need a new `ap_create_secondary_connection()` for now if it's only a pass through `ap_run_create_secondary_connection()`

icing replied:
I think it is good to have and at no cost. We have added wrappers around hooks in the past, when things turned out to need global checks or additional things.

ylavic replied:
Yeah right but if at that time for instance we need more parameters in the helper than in the hook we'd have to create and _ex() version if the helper exists already.
Not a strong opinion though..

icing replied:
Not opposed to it either. Maybe @rpluem can arbitrate...

ylavic commented on server/core.c:
```diff
@@ -5517,6 +5519,55 @@ static conn_rec *core_create_conn(apr_pool_t *ptrans, server_rec *s,
+    apr_pool_tag(pool, "secondary_conn");
+
+    /* first copy everything, then replace what is needed */
+    memcpy(c, master, sizeof(*c));
```
pcalloc + memcpy => pmemdup ?

icing replied:
good point.

icing replied:
Changed as suggested.

Commit:
Merge branch 'trunk' into h2-pollset-impr

Commit:
Added improvements to atomic use of strcut members and made
use of apr_pmemdup on creating secondary connections.

Improved h2_workers.c wake_idle_worker() as suggested.

Commit:
* check allocator creating for OOM and abort when this happens
 * check that c2 creation actually returned a connection
 * make sure out parameter of h2_mplx_worker_c2_done() always
   has a defined value.

ylavic reviewed (approved):
+1

icing commented:
Merged as r1899032 in trunk. Thanks for reviewing!
